Mathlib/Algebra/Star/NonUnitalSubsemiring.lean

```diff
@@ -0,0 +1,233 @@
+/-
+Copyright (c) 2024 Christopher Hoskin. All rights reserved.
+Released under Apache 2. license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Group.Subsemigroup.Basic
+import Mathlib.RingTheory.NonUnitalSubsemiring.Basic
+import Mathlib.Algebra.Ring.Subsemiring.Basic
+import Mathlib.Algebra.Star.Center
+
+/-!
+# Non-unital Star Subsemirings
+
+In this file we define `NonUnitalStarSubsemiring`s and the usual operations on them.
+
+## Implementation
+
+This file is heavily inspired by `Mathlib.Algebra.Star.NonUnitalSubalgebra`.
+
+-/
+
+universe v w w'
+
+variable {A : Type v} {B : Type w} {C : Type w'}
+
+/-- A non-unital star subset is a subset which is closed under the `star`
+operation. -/
+structure StarSubset (A : Type v) [Star A] : Type v where
+  /-- The underlying set of a `SubStar`. -/
+  carrier : Set A
+  /-- The `carrier` of a `StarSubset` is closed under the `star` operation. -/
+  star_mem' : ∀ {a : A} (_ha : a ∈ carrier), star a ∈ carrier
+
+/-- A sub star magma is a subset of a magma which is closed under the `star`-/
+structure SubStarmagma (M : Type v) [Mul M] [Star M] extends Subsemigroup M, StarSubset M : Type v
+
+/-- Reinterpret a `SubStarmagma` as a `Subsemigroup`. -/
+add_decl_doc SubStarmagma.toSubsemigroup
+/-- Reinterpret a `SubStarmagma` as a `StarSubset`. -/
+add_decl_doc SubStarmagma.toStarSubset
+
+/-- A non-unital star subsemiring is a non-unital non-associative ring which is closed under the
+`star` operation. -/
+structure NonUnitalStarSubsemiring (R : Type v) [NonUnitalNonAssocSemiring R] [Star R]
+    extends NonUnitalSubsemiring R, StarSubset R : Type v
+
+/-- Reinterpret a `NonUnitalStarSubsemiring` as a `NonUnitalSubsemiring`. -/
+add_decl_doc NonUnitalStarSubsemiring.toNonUnitalSubsemiring
+/-- Reinterpret a `NonUnitalStarSubsemiring` as a `StarSubset`. -/
+add_decl_doc NonUnitalStarSubsemiring.toStarSubset
+
+/-- A (unital) star subsemiring is a non-associative ring which is closed under the `star`
+operation. -/
+structure StarSubsemiring (R : Type v) [NonAssocSemiring R] [Star R]
```

Comment:
It's weird that this is in the `NonUnitalSubsemiring` file. Perhaps you can move the unital stuff out of here into a separate file (and possibly even a separate PR).